### PR TITLE
Remove module globals IPC channel, return struct directly

### PR DESCRIPTION
**Pull request by mcereal**
## Summary

Eliminate the `.module_globals_*` file IPC channel by converting `lower_module_bindings_to_globals` and its helper `_process_one_binding` to return structured results instead of writing to scratch files. This removes 13 `fs.writeFile` calls and 6 `fs.readFile` calls, eliminates an O(N²) read-modify-write pattern in the append helper, and makes both functions pure.

## Key Changes

- **New struct `BindingLoweringResult`** in `llvm/types.sfn`: wraps `preamble_lines`, `locals`, and `needs_init` flag to carry per-binding lowering results out of the helper function.

- **`_process_one_binding` refactored**: now returns `BindingLoweringResult` at all 4 early-exit points and 1 final return instead of writing `.module_globals_preamble` and `.module_globals_locals` files. Removed `![io]` effect annotation — function is now pure.

- **`lower_module_bindings_to_globals` refactored**: accumulates per-binding results in the outer loop using `.push()` on arrays, fully populates `ModuleGlobalLoweringResult` at its single return point, and removes all `fs.writeFile` calls. Removed `![io]` effect annotation.

- **Caller in `lowering_core.sfn`** simplified: replaces six `fs.readFile` + `split_lines_local` + `_parse_module_globals_locals` calls with direct struct field reads from the returned `ModuleGlobalLoweringResult`.

- **Dead code removal**: `_parse_module_globals_locals` (40 lines) deleted from `lowering_text_utils.sfn` as it is no longer needed. `_append_to_file` helper deleted entirely.

- **Legacy cleanup**: `.module_globals_` prefix retained in `cli_commands._clean_lowering_state` with a comment explaining it cleans stale files from older build directories.

## Implementation Details

- The per-binding helper previously performed O(N²) filesystem operations via `_append_to_file`, which read the entire accumulating file, appended a line, and wrote it back on each call. This is now eliminated.
- Both functions transition from file-IPC stubs to proper implementations that populate their return structures.
- The early-return path for empty bindings is preserved.
- Three `![io]` effect annotations are narrowed: the helper and writer become pure, and the append helper is deleted entirely.

https://claude.ai/code/session_01ASZvPexEw6koHyc5mKGn4q

### CODE REVIEW

File: compiler/src/llvm/lowering/module_globals.sfn [hunk 1]
```diff
@@ -170,44 +162,37 @@ fn _process_one_binding(binding: NativeBinding, context: TypeContext, scope_id:
...
             }
-            if !is_const { needs_init_result = "init"; }
+            if !is_const { needs_init = true; }
         }
-        _append_to_file("build/sailfin/.module_globals_preamble", global_name
-            + " = internal global "
-            + llvm_type
+        preamble_lines.push(global_name + " = internal global " + llvm_type
             + " "
             + initializer_text);
```
[the remark this answers is not in the document]
Comment (mcereal) [reply compiler/src/llvm/lowering/module_globals.sfn:169]: This is a pre-existing latent issue, not introduced by this PR. The old file-IPC path behaved identically: `_process_one_binding` set `needs_init_result = "init"` for non-const initializers, the loop-exit wrote `.module_globals_needs_init` as `"1"`, and the reader in `lowering_core.sfn:651` propagated that to `module_globals_needs_init_call`. `init_function_lines` was also always `[]` pre-PR — `.module_globals_init` was initialized empty and nothing ever appended to it. Behavior on both the const and non-const paths is byte-identical before and after this PR.

The path is also not exercised by self-hosting: the only compiler module with a top-level `let` is `compiler/src/version.sfn`, whose `__version_fallback__ = "0.5.7"` hits the string-literal fast path (`string_handled = true` → `needs_init` stays false). So the dangling-init-symbol case you describe is dead code during selfhost.

Fixing the latent case (either generating the init function body or diagnosticing unsupported initializers) is out of scope for this channel-removal PR — it would conflate a behavior change with the IPC refactor. Happy to open a follow-up issue to track it; feel free to leave this thread unresolved as a marker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: compiler/src/llvm/lowering/module_globals.sfn [hunk 2]
```diff
@@ -170,44 +162,37 @@ fn _process_one_binding(binding: NativeBinding, context: TypeContext, scope_id:
...
+        locals: locals,
+        needs_init: needs_init
+    };
 }
 
-fn lower_module_bindings_to_globals(bindings: NativeBinding[], context: TypeContext, module_name: string) -> ModuleGlobalLoweringResult ![io] {
+fn lower_module_bindings_to_globals(bindings: NativeBinding[], context: TypeContext, module_name: string) -> ModuleGlobalLoweringResult {
     let init_symbol = module_init_symbol(module_name);
 
     if bindings.length == 0 {
```
[the remark this answers is not in the document]
Comment (mcereal) [reply compiler/src/llvm/lowering/module_globals.sfn:195]: `make rebuild` self-hosting is the regression contract for this procedure (see [docs/build-performance.md §267 step 7](../blob/main/docs/build-performance.md#procedure-removing-an-ipc-channel)): if the in-memory path is incomplete, the rebuild fails with a resolution error. This matches the precedent from PR #178 / PR #183 / PR #184 — each removed one IPC channel without adding a dedicated test capsule, and each relied on self-host + the existing suite for regression coverage.

The code path is exercised: `compiler/src/version.sfn` has a top-level `let __version_fallback__ = "0.5.7";`, so `lower_module_bindings_to_globals` runs with `bindings.length > 0` during selfhost. A rebuild from a prior seed into the new binary that then successfully builds itself (and runs `make test`) is end-to-end validation that the preamble lines, locals, and init symbol all get through the struct-return path correctly.

Adding a dedicated test capsule here would duplicate what selfhost already gates, and would constrain the allowed shape of future refactors (a test capsule that pinned specific `@global.*` LLVM lines would need updating every time emission changes). The project convention for this phase is to lean on the selfhost + suite combo rather than per-channel unit tests.

Awaiting CI to confirm both linux-x86_64 and macos-arm64 pass.

---
_Generated by [Claude Code](https://claude.ai/code)_
